Verify transactions are well formed when computing the block's data hash instead of externally

yacovm opened the pull request:
To make sure the function VerifyTransactionsAreWellFormed is called in the right places,
I moved its invocation to the computation of the Fabric's block data hash.

yacovm committed:
Verify transactions are well formed when computing the block's data hash instead of externally

To make sure the function VerifyTransactionsAreWellFormed is called in the right places,
I moved its invocation to the computation of the Fabric's block data hash.

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

MayRosenbaum committed:
verify tx are well formed - fixed

Signed-off-by: May Rosenbaum <mayro1595@gmail.com>

yacovm committed:
Compute the Block's data hash directly when proposing blocks

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

tock-ibm commented on orderer/consensus/smartbft/assembler.go:
```diff
@@ -51,7 +53,11 @@ func (a *Assembler) AssembleProposal(metadata []byte, requests [][]byte) (nextPr
+	// Instead of calling BlockDataHash() which calls VerifyTransactionsAreWellFormed(),
+	// we directly compute the block's data hash.
+	dataHash := sha256.Sum256(bytes.Join(block.Data.Data, nil))
+	block.Header.DataHash = dataHash[:]
```
I don't like this... I think the mix of calling BlockDataHash and direct computation like here is s recipe for trouble. Especially if we go for an update of said computation in V3. 

It would be much better in my opinion to offer two  versions of the original BlockDataHash function, one that does verify and one that does not. 
You can change the name / signature and this way we'll catch all the instances. For example:

BlockDataHash(...) ([]byte, error) \\ with verify

BlockDataHashCompute (...) ([]byte) \\ w/o verify

or maybe add a bool to the BlockDataHash (bc, verfyTxs bool) that selects whether to verify or not.

or finally, bite the bullet and verify even if you don't need to.

tock-ibm commented on common/ledger/blockledger/util.go:
```diff
@@ -76,7 +76,7 @@ func CreateNextBlock(rl Reader, messages []*cb.Envelope) *cb.Block {
 
 	block := protoutil.NewBlock(nextBlockNumber, previousBlockHash)
-	block.Header.DataHash = protoutil.BlockDataHash(data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(data)
```
Here you use the version that verifies the TXs even though it is really unnecessary. In another place below you use the raw computation... I think a uniform approach is better. See comment bellow.

tock-ibm commented:
I have to admit that I am not entirely comfortable with this approach. There are a lot of places where the TXs verification is not necessary, and so the error is just ignored... so why not replace all the original instances of BlockDataHash with two versions: one that does not verify, and one that does? reviewing such a change is exactly like reviewing this one, and having to decide whether it is OK not to check the error or it is a must. Plus, it avoids checking the TXs in places where it is really not needed.

yacovm committed:
Address code review comments

Move all instances where we ignore the error or compute directly the hash into a separate function.

Signed-off-by: Yacov Manevich <yacov.manevich@ibm.com>

yacovm commented:
> I have to admit that I am not entirely comfortable with this approach. There are a lot of places where the TXs verification is not necessary, and so the error is just ignored... so why not replace all the original instances of BlockDataHash with two versions: one that does not verify, and one that does? reviewing such a change is exactly like reviewing this one, and having to decide whether it is OK not to check the error or it is a must. Plus, it avoids checking the TXs in places where it is really not needed.

Done.

tock-ibm commented on cmd/osnadmin/main_test.go:
```diff
@@ -820,7 +820,7 @@ func blockWithGroups(groups map[string]*cb.ConfigGroup, channelID string) *cb.Bl
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Why not ComputeBlockDataHash?

yacovm replied:
Why does it matter how I do it? the result is the same, and it's a test.

tock-ibm commented on common/ledger/testutil/test_helper.go:
```diff
@@ -303,7 +307,7 @@ func NewBlock(env []*common.Envelope, blockNum uint64, previousHash []byte) *com
 		block.Data.Data = append(block.Data.Data, txEnvBytes)
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Why not ComputeBlockDataHash?

yacovm replied:
Why does it matter how I do it? the result is the same, and it's a test.

tock-ibm commented on core/committer/txvalidator/v14/txvalidator_test.go:
```diff
@@ -277,9 +277,10 @@ func TestTxValidationFailure_InvalidTxid(t *testing.T) {
 		},
 	}
 
+	hash, _ := protoutil.BlockDataHash(block.Data)
```
Why not ComputeBlockDataHash? or check the error?

yacovm replied:
ok checking the error is a fair point. but why does it matter how I do it? the result is the same, and it's a test.

tock-ibm commented on core/committer/txvalidator/v20/txvalidator_test.go:
```diff
@@ -231,9 +231,10 @@ func TestTxValidationFailure_InvalidTxid(t *testing.T) {
 		},
 	}
 
+	hash, _ := protoutil.BlockDataHash(block.Data)
```
Why not ComputeBlockDataHash? or check the error?

yacovm replied:
ok checking the error is a fair point. but why does it matter how I do it? the result is the same, and it's a test.

tock-ibm commented on core/ledger/kvledger/benchmark/chainmgmt/block_gen.go:
```diff
@@ -82,7 +82,7 @@ func (bg *blkGenerator) nextBlock() *common.Block {
 		return nil
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Why not ComputeBlockDataHash?

tock-ibm commented on core/ledger/kvledger/tests/committer.go:
```diff
@@ -90,7 +90,7 @@ func (g *blkGenerator) nextBlockAndPvtdata(trans []*txAndPvtdata, missingPvtData
 		}
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Why not ComputeBlockDataHash?

tock-ibm commented on integration/raft/channel_participation_test.go:
```diff
@@ -1081,7 +1081,7 @@ func createJoinBlockDefineSystemChannel(channelID string) *common.Block {
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Why not ComputeBlockDataHash?

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -188,7 +188,7 @@ var _ = Describe("EndToEnd reconfiguration and onboarding", func() {
 			genesisBlock.Data.Data[0], err = protoutil.Marshal(envelope)
 			Expect(err).NotTo(HaveOccurred())
-			genesisBlock.Header.DataHash = protoutil.BlockDataHash(genesisBlock.Data)
+			genesisBlock.Header.DataHash, _ = protoutil.BlockDataHash(genesisBlock.Data)
```
Why not ComputeBlockDataHash?

tock-ibm commented on integration/raft/config_test.go:
```diff
@@ -2127,7 +2127,7 @@ func configToGenesisBlock(configBlock *common.Block, channelId string) *common.B
 	Expect(err).NotTo(HaveOccurred())
 
-	genesisBlock.Header.DataHash = protoutil.BlockDataHash(genesisBlock.Data)
+	genesisBlock.Header.DataHash, _ = protoutil.BlockDataHash(genesisBlock.Data)
```
Why not ComputeBlockDataHash?

tock-ibm commented on integration/smartbft/smartbft_test.go:
```diff
@@ -1988,7 +1989,7 @@ func createPrePrepareRequest(
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
Why not ComputeBlockDataHash?

tock-ibm commented on internal/ledgerutil/verify/verify.go:
```diff
@@ -229,8 +229,12 @@ func checkBlock(block *common.Block, previousHash []byte) (*blockCheckResult, er
+	hash, err := protoutil.BlockDataHash(block.Data)
+	if err != nil {
+		return nil, err
+	}
```
Why not add to the results? this is a utility to verify existing ledgers, and discovering such an error should be reported in the tool output, not crash the tool.

tock-ibm replied:
In addition, a test case with such an error should be added to the unit test of 
`func TestVerify(t *testing.T)`  of verify_test.go

tock-ibm commented on internal/peer/gossip/mcs_test.go:
```diff
@@ -475,7 +475,7 @@ func mockBlock(t *testing.T, channel string, seqNum uint64, localSigner *mocks.S
 		block.Header.DataHash = dataHash
 	} else {
-		block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+		block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on internal/peer/gossip/mcs_test.go:
```diff
@@ -518,7 +518,7 @@ func mockBlockBFT(t *testing.T, channel string, seqNum uint64, localSigner *mock
 		block.Header.DataHash = dataHash
 	} else {
-		block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+		block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/channelparticipation/validator.go:
```diff
@@ -31,7 +31,11 @@ func ValidateJoinBlock(configBlock *cb.Block) (channelID string, err error) {
+	dataHash, err := protoutil.BlockDataHash(configBlock.Data)
+	if err != nil {
+		return "", err
+	}
```
Add a unit test for when that happens, please

yacovm replied:
I removed it, because we only get config blocks so we cannot fail this check.

tock-ibm commented on orderer/common/channelparticipation/validator_test.go:
```diff
@@ -162,7 +162,7 @@ func blockWithGroups(groups map[string]*cb.ConfigGroup, channelID string) *cb.Bl
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/channelparticipation/validator_test.go:
```diff
@@ -183,7 +183,7 @@ func nonConfigBlock() *cb.Block {
 		},
 	}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/cluster/util.go:
```diff
@@ -260,11 +260,11 @@ func VerifyBlockHash(indexInBuffer int, blockBuff []*common.Block) error {
-	if err := protoutil.VerifyTransactionsAreWellFormed(block); err != nil {
+	seq := block.Header.Number
+	dataHash, err := protoutil.BlockDataHash(block.Data)
+	if err != nil {
```
As mentioned in the previous PR, this will crash when `block.Data==nil`, see comment below.

tock-ibm commented on protoutil/blockutils.go:
```diff
@@ -300,18 +307,18 @@ func searchConsenterIdentityByID(consenters []*cb.Consenter, identifier uint32)
-func VerifyTransactionsAreWellFormed(block *cb.Block) error {
-	if block == nil || block.Data == nil || len(block.Data.Data) == 0 {
+func VerifyTransactionsAreWellFormed(bd *cb.BlockData) error {
+	if bd.Data == nil || len(bd.Data) == 0 {
```
you also have to check against `bd == nil`

tock-ibm commented on protoutil/blockutils_test.go:
```diff
@@ -36,15 +38,17 @@ func TestNewBlock(t *testing.T) {
-	block.GetHeader().DataHash = protoutil.BlockDataHash(data)
+	block.GetHeader().DataHash, _ = protoutil.BlockDataHash(data)
+
+	dataHash, _ := protoutil.BlockDataHash(data)
```
check no error?
Is there a test case that returns error?
Check when `data` is nil?

tock-ibm commented on protoutil/blockutils_test.go:
```diff
@@ -606,11 +610,15 @@ func TestVerifyTransactionsAreWellFormed(t *testing.T) {
-			if tst.expectedError == "" {
-				require.NoError(t, err)
+			if tst.block == nil || tst.block.Data == nil {
+				require.Error(t, errors.New("empty block"))
```
what does this check? that an error is an error ;-)

tock-ibm commented on orderer/common/cluster/util_test.go:
```diff
@@ -194,7 +194,7 @@ func TestVerifyBlockBFT(t *testing.T) {
 	block.Data.Data[0] = protoutil.MarshalOrPanic(configTx)
 	block.Header.Number = 2
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/cluster/util_test.go:
```diff
@@ -334,7 +334,7 @@ func createBlockChain(start, end uint64) []*common.Block {
 	for seq := start; seq <= end; seq++ {
 		block := newBlock(seq)
-		block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+		block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/follower/follower_chain_test.go:
```diff
@@ -945,11 +945,11 @@ func (mbc *memoryBlockChain) fill(numBlocks uint64) {
 		if i == 0 {
 			block = makeConfigBlock(i, prevHash, 0)
-			block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+			block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/follower/follower_chain_test.go:
```diff
@@ -945,11 +945,11 @@ func (mbc *memoryBlockChain) fill(numBlocks uint64) {
 			block = protoutil.NewBlock(i, prevHash)
 			block.Data.Data = [][]byte{{uint8(i)}, {uint8(i)}}
-			block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+			block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/follower/follower_chain_test.go:
```diff
@@ -963,7 +963,7 @@ func (mbc *memoryBlockChain) appendConfig(isMember uint8) {
 	h := uint64(len(mbc.chain))
 	configBlock := makeConfigBlock(h, protoutil.BlockHeaderHash(mbc.chain[h-1].Header), isMember)
-	configBlock.Header.DataHash = protoutil.BlockDataHash(configBlock.Data)
+	configBlock.Header.DataHash, _ = protoutil.BlockDataHash(configBlock.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/follower/follower_chain_test.go:
```diff
@@ -998,7 +998,7 @@ func makeConfigBlock(num uint64, prevHash []byte, isMember uint8) *common.Block
 	}
 	block.Data.Data = append(block.Data.Data, protoutil.MarshalOrPanic(env))
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/common/multichannel/blockwriter_test.go:
```diff
@@ -62,8 +64,14 @@ func TestCreateBlock(t *testing.T) {
 		{Payload: []byte("some other bytes")},
 	})
 
+	dataHash := sha256.Sum256(bytes.Join(block.Data.Data, nil))
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/consensus/etcdraft/blockcreator_test.go:
```diff
@@ -25,12 +25,14 @@ func TestCreateNextBlock(t *testing.T) {
 	}
 
 	second := bc.createNextBlock([]*cb.Envelope{{Payload: []byte("some other bytes")}})
+	hash := protoutil.ComputeBlockDataHash(second.Data)
```
you can save a line here

tock-ibm commented on orderer/consensus/etcdraft/blockcreator_test.go:
```diff
@@ -25,12 +25,14 @@ func TestCreateNextBlock(t *testing.T) {
 	require.Equal(t, protoutil.BlockHeaderHash(first.Header), second.Header.PreviousHash)
 
 	third := bc.createNextBlock([]*cb.Envelope{{Payload: []byte("some other bytes")}})
+	hash = protoutil.ComputeBlockDataHash(third.Data)
```
you can save a line here

tock-ibm commented on orderer/consensus/smartbft/assembler_test.go:
```diff
@@ -161,7 +161,7 @@ func makeConfigBlock(seq uint64) *common.Block {
 	block := protoutil.NewBlock(seq, nil)
 	block.Data = &common.BlockData{Data: requests}
-	block.Header.DataHash = protoutil.BlockDataHash(block.Data)
+	block.Header.DataHash, _ = protoutil.BlockDataHash(block.Data)
```
why not ComputeBlockDataHash?

tock-ibm commented on orderer/consensus/smartbft/verifier.go:
```diff
@@ -237,14 +237,14 @@ func verifyHashChain(block *cb.Block, prevHeaderHash string) error {
 	}
 
-	if err := protoutil.VerifyTransactionsAreWellFormed(block); err != nil {
+	dataHashString := hex.EncodeToString(block.Header.DataHash)
```
it is probably better to compute that after the `if``

tock-ibm commented on orderer/common/channelparticipation/validator.go:
```diff
@@ -31,10 +29,6 @@ func ValidateJoinBlock(configBlock *cb.Block) (channelID string, err error) {
-	if !bytes.Equal(protoutil.BlockDataHash(configBlock.Data), configBlock.Header.DataHash) {
-		return "", errors.New("invalid block: Header.DataHash is different from Hash(block.Data)")
-	}
-
```
Why did you remove this test completely and not replace it with the new call? I agree that well formedness does not need to be checked after protoutil.IsConfigBlock but what abouts data hash vs the header.DataHash?

e.g.
```
if !bytes.Equal(protoutil.ComputeBlockDataHash(configBlock.Data), configBlock.Header.DataHash) {
		return "", errors.New("invalid block: Header.DataHash is different from Hash(block.Data)")
	}
```

yacovm replied:
You're right, my bad.

tock-ibm commented on orderer/common/channelparticipation/validator_test.go:
```diff
@@ -39,16 +39,6 @@ func TestValidateJoinBlock(t *testing.T) {
-			}(),
-			expectedChannelID: "",
-			expectedErr:       errors.New("invalid block: Header.DataHash is different from Hash(block.Data)"),
-		},
```
Same question as above, I think this use case should be tested and protected against.

yacovm replied:
done
